api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -495,4 +496,89 @@ public Workspace setPublished(
 
     return this.saveWithLastModified(workspace);
   }
+
+  @Override
+  public List<UserRecentWorkspace> getRecentWorkspacesByUser(long userId) {
+    List<UserRecentWorkspace> userRecentWorkspaces =
+        userRecentWorkspaceDao.findByUserIdOrderByLastAccessDateDesc(userId);
+    enforceFirecloudAclsInRecentWorkspaces(userRecentWorkspaces);
+    return userRecentWorkspaceDao.findByUserIdOrderByLastAccessDateDesc(userId);
+  }
+
+  private void enforceFirecloudAclsInRecentWorkspaces(List<UserRecentWorkspace> recentWorkspaces) {
+    // Map recentWorkspaces by id because we'll want to access them like that later when we delete
+    Map<Long, UserRecentWorkspace> recentWorkspacesById =
+        recentWorkspaces.stream()
+            .collect(Collectors.toMap(UserRecentWorkspace::getWorkspaceId, rw -> rw));
+
+    // Grab the dbWorkspaces.
+    List<Workspace> dbWorkspaces =
+        workspaceDao.findAllByWorkspaceIdIn(
+            recentWorkspaces.stream()
+                .map(UserRecentWorkspace::getWorkspaceId)
+                .collect(Collectors.toList()));
+
+    // Map by primary key id bc it's illegal to string streams along in Java. not pretty enough.
+    Map<Long, Workspace> dbWorkspacesById =
+        dbWorkspaces.stream()
+            .collect(Collectors.toMap(Workspace::getWorkspaceId, dbWorkspace -> dbWorkspace));
+
+    // Unfortunately it doesn't look like there's a way to batch call this currently. At least we
+    // can make it parallel.
```

Comment:
Is this already parallel or is that the goal? From what I can tell `forEach` is sequential and single-threaded.